app/services/slot_builder.rb

```diff
@@ -0,0 +1,175 @@
+      return range.begin..busy_time.starts_at if range.cover?(busy_time.starts_at)
+      return busy_time.ends_at..range.end if range.cover?(busy_time.ends_at)
+    end
+
+    def slots_for(plage_ouverture_free_times, motif)
+      slots = []
+      plage_ouverture_free_times.each do |plage_ouverture, free_times|
+        free_times.each do |free_time|
+          slots += calculate_slots(free_time, motif, plage_ouverture)
+        end
+      end
+      slots
+    end
+
+    def calculate_slots(free_time, motif, plage_ouverture)
+      slots = []
+      possible_slot_time = free_time.begin..(free_time.begin + motif.default_duration_in_min.minutes)
+      while possible_slot_time.end <= free_time.end
+        slots << Creneau.new(
+          starts_at: possible_slot_time.begin,
+          motif: motif,
+          lieu_id: plage_ouverture.lieu_id,
+          agent: plage_ouverture.agent
+        )
+        possible_slot_time = possible_slot_time.end..(possible_slot_time.end + motif.default_duration_in_min.minutes)
+      end
+      slots
+    end
+  end
+
+  class BusyTime
+    attr_reader :starts_at, :ends_at
+
+    def initialize(object)
+      case object
+      when Date
+        @starts_at = object.beginning_of_day
+        @ends_at = object.end_of_day
+      when Rdv, Recurrence::Occurrence
+        @starts_at = object.starts_at
+        @ends_at = object.ends_at
+      when Absence
+        @starts_at = object.start_time.on(object.first_day)
+        @ends_at = object.end_time.on(object.end_day.presence || object.first_day)
+      else
+        raise ArgumentError, "busytime can't be build with a #{object.class}"
+      end
+    end
+
+    def range
+      (starts_at..ends_at)
+    end
+
+    class << self
+      def busy_times_for(range, plage_ouverture, off_days = [])
+        # c'est là que l'on execute le SQL
+        # TODO : Peut-être cacher la récupération de l'ensemble des RDV et absences concernées (pour n'avoir que deux requêtes) puis faire des selections dessus pour le filtre sur le range
+
+        busy_times = busy_times_from_rdvs(range, plage_ouverture)
+        busy_times += busy_times_from_absences(range, plage_ouverture)
+        busy_times += busy_times_from_off_days(off_days)
+
+        # Le tri est nécessaire, surtout pour les surcharges.
+        busy_times.sort_by(&:starts_at)
+      end
+
+      def busy_times_from_rdvs(range, plage_ouverture)
+        plage_ouverture_starts_in_range = plage_ouverture.agent.rdvs.not_cancelled.where(starts_at: range)
+        plage_ouverture_ends_in_range = plage_ouverture.agent.rdvs.not_cancelled.where(ends_at: range)
+        plage_ouverture_starts_in_range.or(plage_ouverture_ends_in_range).map do |rdv|
+          BusyTime.new(rdv)
+        end
+      end
+
+      def busy_times_from_absences(range, plage_ouverture)
+        absences = plage_ouverture.agent.absences.where(organisation: plage_ouverture.organisation).in_range(range)
```

Comment:
Il faut un not_expired ici, suite à #1899

(Ici aussi, suggestion de style pour faire une chose par ligne)
```suggestion
        absences = plage_ouverture.agent.absences
          .where(organisation: plage_ouverture.organisation)
          .in_range(range)
```

Reply:
Tu aurais pu en profiter pour ajouter le `not_expired` :stuck_out_tongue: